### ISSUE TITLE
Images & Documents Search - Search Query should not activate on empty searches

### ISSUE THREAD

**Issue by lb-** [wagtail/wagtail#3997]
Found a bug? Please fill out the sections below. 👍

### Issue Summary

When navigating through collections of Images, the search field (which is focused by default) is very sensitive to empty search changes (such as pressing the `down key`) which then clear the current page and collection you are on.

### Steps to Reproduce

It's essential that you provide enough information for someone else to replicate the problem you're seeing. Simply describing something that's broken on your current project is not enough!

1. Load up the wagtaildemo
2. Go to Admin > Settings > Collection & Click 'Add a Collection' (create one - can be any name)
3. Add 30 images, 15 in the 'Root' Collection and 15 in the new connection you just created.
4. Go to the images page `/admin/images/`
5. Select the non-root collection, URL should update to something like `/admin/images/?collection_id=2`
6. Scroll to the bottom of the page and click 'Next' to go to page two. URL should update to something like `/admin/images/?collection_id=2&p=2` (Note: This step is not needed, just explains workflow).
7. Click the down arrow on your keyboard to move down the page.
7a. **Expected**: Either nothing (because the search box is in focus) or page moves down
7b. **Actual**: A search is triggered in the search field, this causes a search for 'nothing' which reloads the URL to `/admin/images/?q=`

The main issue here is now the user is very confused, they were on page two of a specific collection and now they suddenly ended up back at the home page of the images?

This also happens on Documents.

### Technical details

* Python version: 2.7.12
* Django version: Django v.1.11.6
* Wagtail version: 1.13
* Browser version: Chrome 61

### Potential Solution (basic)

It appears that the event `keyup cut paste` is causing this search trigger here.
`$(window.headerSearch.termInput).on('keyup cut paste', function() {...`
[wagtailadmin/static_src/wagtailadmin/js/core.js#L222](https://github.com/wagtail/wagtail/blob/7a1455551bc83a60def638b3351a641035a95d64/wagtail/wagtailadmin/static_src/wagtailadmin/js/core.js#L222)

Maybe in the search function we could add a check for empty values?

```
function search() {
    var searchInputValue = $(window.headerSearch.termInput).val();
    // if searchInputValue is not null, or contains more than an empty string.
    if (searchInputValue) {
    ... // rest of the function that exists already
    }
}
```

I am sure there is a more elegant javascript way to do this but it will mean we won't get false triggers.

### Potential Solution (advanced)

Along with ignoring the query if it is blank, we should consider the currently selected collection which is cleared whenever a new search takes place.

ie. URL should become

`/admin/images/?collection_id=2&q=my+query`

NOT

`/admin/images/?q=my+query`

Along with considering the current search query when the collection is changed.

However, these are probably two extra issues that need to be created.

**Comment by lb-** [comment]
@jjanssen I would like your thoughts on this proposed way of doing the search, essentially the search only runs if it has changed (ignoring trimmed whitespace). Also, do we have a way to do tests with the JS code?

This would replace the lines:
https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailadmin/static_src/wagtailadmin/js/core.js#L217-L253

```
if (window.headerSearch) {
        var searchCurrentIndex = 0;
        var searchNextIndex = 0;

        $(window.headerSearch.termInput).on('keyup cut paste', function() {
            clearTimeout($.data(this, 'timer'));
            var wait = setTimeout(search, 200);
            $(this).data('timer', wait);
        });

        // auto focus on search box
        $(window.headerSearch.termInput).trigger('focus');

        function search() {
            var workingClasses = 'icon-spinner';

            var newQuery = $(window.headerSearch.termInput).val();
            var currentQuery = getURLParam('q');
            if (! currentQuery) { currentQuery = ""; }
            // only do the query if it has changed for trimmed queries
            // eg. " " === "" and "firstword " ==== "firstword"
            if (currentQuery.trim() !== newQuery.trim()) {
                $(window.headerSearch.termInput).parent().addClass(workingClasses);
                searchNextIndex++;
                var index = searchNextIndex;
                $.ajax({
                    url: window.headerSearch.url,
                    data: {q: newQuery},
                    success: function(data, status) {
                        if (index > searchCurrentIndex) {
                            searchCurrentIndex = index;
                            $(window.headerSearch.targetOutput).html(data).slideDown(800);
                            window.history.pushState(null, 'Search results', '?q=' + newQuery);
                        }
                    },
                    complete: function() {
                        $(window.headerSearch.termInput).parent().removeClass(workingClasses);
                      }
                });
            }
      };

      getURLParam = function(name) {
          var results = new RegExp('[\?&]' + name + '=([^]*)').exec(window.location.href);
          if (results == null) {
              return null;
          } else {
              return results[1] || null;
          }
      };

    }
```

**Comment by jjanssen** [comment]
Hi @lb- this all seems pretty straight-forward. 

My suggestion would be to let the `getURLParam` return an empty string instead of `null` since you're already overriding this with `if (! currentQuery) { currentQuery = ""; }` after you requested it one line before.

Maybe we can write the return a little cleaner rather then having an explicit if-else with the `null`-check. Unless there is a valid reason for this of course. If we need to check it would be better to check it [strictly](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators#Identity_strict_equality_()). 

```
getURLParam = function() {
    var results = new RegExp('[\?&]' + name + '=([^]*)').exec(window.location.href);

    if (results) {
        return results[1] || '';
    }

    return '';
}
```

For `window.location.href` we can use `window.location.search` since this will contain our query parameters. For example its output is like: `?page=0&sort=lastName%2Casc`. Less characters to run through our regular expression 🙌 

As far as testing is concerned. Within this section of the codebase we unfortunately do not have tests for it. When introducing the react-explorer Thibaud and I implemented [Jest](https://facebook.github.io/jest/) for the `./client` scripting in the root. Maybe it is something we can extend to our existing codebase, but we have not looked into yet.

**Comment by lb-** [comment]
@jjanssen thanks for that epic feedback, about to do a PR, would love any additional input when you have a min.

**Comment by jjanssen** [comment]
Closing this since it was merged with 7c54c14
